daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractKeySpec.scala

```diff
@@ -0,0 +1,430 @@
+                (Some[Ref.Name]("receiver"), ValueParty(bob)),
+              ),
+            ),
+            "",
+          )
+        ),
+      toContractId("BasicTests:WithKey:1") ->
+        withKeyContractInst,
+    )
+
+  val defaultKey = Map(
+    GlobalKey.assertBuild(
+      TypeConName(basicTestsPkgId, withKeyTemplate),
+      ValueRecord(None, ImmArray((None, ValueParty(alice)), (None, ValueInt64(42)))),
+    )
+      ->
+        toContractId("BasicTests:WithKey:1")
+  )
+
+  private[this] def lookupPackage(pkgId: PackageId): Option[Package] = {
+    allPackages.get(pkgId)
+  }
+
+  private[this] def lookupKey(key: GlobalKeyWithMaintainers): Option[ContractId] =
+    (key.globalKey.templateId, key.globalKey.key) match {
+      case (
+            BasicTests_WithKey,
+            ValueRecord(_, ImmArray((_, ValueParty(`alice`)), (_, ValueInt64(42)))),
+          ) =>
+        Some(toContractId("BasicTests:WithKey:1"))
+      case _ =>
+        None
+    }
+
+  private[this] val suffixLenientEngine = newEngine()
+  private[this] val preprocessor =
+    new preprocessing.Preprocessor(
+      ConcurrentCompiledPackages(suffixLenientEngine.config.getCompilerConfig)
+    )
+
+  "contract key" should {
+    val now = Time.Timestamp.now()
+    val submissionSeed = crypto.Hash.hashPrivateKey("contract key")
+    val txSeed = crypto.Hash.deriveTransactionSeed(submissionSeed, participant, now)
+
+    // TEST_EVIDENCE: Semantics: contract keys should be evaluated only when executing create
+    "be evaluated only when executing create" in {
+      val templateId =
+        Identifier(basicTestsPkgId, "BasicTests:ComputeContractKeyWhenExecutingCreate")
+      val createArg =
+        ValueRecord(
+          Some(templateId),
+          ImmArray((Some[Name]("owner"), ValueParty(alice))),
+        )
+      val exerciseArg =
+        ValueRecord(
+          Some(Identifier(basicTestsPkgId, "BasicTests:DontExecuteCreate")),
+          ImmArray.Empty,
+        )
+
+      val submitters = Set(alice)
+
+      val Right(cmds) = preprocessor
+        .preprocessCommands(
+          ImmArray(
+            CreateAndExerciseCommand(templateId, createArg, "DontExecuteCreate", exerciseArg)
+          )
+        )
+        .consume(_ => None, lookupPackage, lookupKey)
+
+      val result = suffixLenientEngine
+        .interpretCommands(
+          validating = false,
+          submitters = submitters,
+          readAs = Set.empty,
+          commands = cmds,
+          ledgerTime = now,
+          submissionTime = now,
+          seeding = InitialSeeding.TransactionSeed(txSeed),
+        )
+        .consume(_ => None, lookupPackage, lookupKey)
+      result shouldBe a[Right[_, _]]
+    }
+
+    // TEST_EVIDENCE: Semantics: contract keys should be evaluated after ensure clause
+    "be evaluated after ensure clause" in {
+      val templateId =
+        Identifier(basicTestsPkgId, "BasicTests:ComputeContractKeyAfterEnsureClause")
```

Comment:
I’m confused. What is this test testing that the one above doesn’t test?
`ComputeContractKeyWhenExecutingCreate` just calls `ComputeContractKeyAfterEnsureClause`.

Reply:
It looks like the first test shows the behaviour when the `create` is not actually executed:
```
        let _ignore = create ComputeContractKeyAfterEnsureClause with owner
```
I'm not sure how useful it is to test that here, as that's just standard behaviour for a action which is not executed,